homeassistant/components/stream/worker.py

```diff
@@ -98,19 +110,32 @@ def reset(self, video_pts):
                 template=self._input_audio_stream
             )
 
-    def mux_packet(self, packet):
+    def mux_packet(self, packet: av.Packet) -> None:
         """Mux a packet to the appropriate output stream."""
 
         # Check for end of segment
-        if packet.stream == self._input_video_stream and packet.is_keyframe:
-            duration = (packet.pts - self._segment_start_pts) * packet.time_base
-            if duration >= MIN_SEGMENT_DURATION:
-                # Save segment to outputs
-                self.flush(duration)
+        if packet.stream == self._input_video_stream:
+
+            self.check_flush_part(packet)
+
+            if (
+                packet.is_keyframe
+                and (
+                    segment_duration := (packet.pts - self._segment_start_pts)
+                    * packet.time_base
+                )
+                >= MIN_SEGMENT_DURATION
+            ):
+                # Flush segment (also flushes the stub part segment)
+                self.flush(segment_duration, packet)
```

Comment:
I think there may be a way to structure `check_flush_part` and `flush` so that the interactions between them are a bit more connected, but i'm not sure yet. I needed to take down [my own notes](https://docs.google.com/document/d/1c1adh-pBBnp4Wn9ZEptoQxvpM1SdhHfD80n7qnkVa5k/edit?usp=sharing) to try to understand the state machine a little better first.

Reply:
The reason there is an extra `Part` that is appended in `flush` is related to the nuance I mentioned above regarding the 1 packet delay between the packet we have muxed and what exists in the most recent fragment. Because of this there will always be a stub section at the end of a segment which has not been included in previous finished fragments, and it is only written when we call `_av_output.close()` (this stub may be as short as 1 packet or it may be as long as all the other fragments).

Reply:
The other angle i'm thinking about his if you unroll this all into a single function: https://docs.google.com/document/d/1c1adh-pBBnp4Wn9ZEptoQxvpM1SdhHfD80n7qnkVa5k/edit#bookmark=id.oyigdot10sew does it get simpler at all. (that isn't correct, but i feel like it will help me understand why logic is different for a Part depending on the scenarion.

Reply:
See above